registry.go

```diff
@@ -64,10 +64,22 @@ func Logout(ctx context.Context, store Store, registryName string) error {
 	return nil
 }
 
+// Credential returns a Credential() function that can be used by auth.Client.
+func Credential(store Store) func(context.Context, string) (auth.Credential, error) {
+	return func(ctx context.Context, reg string) (auth.Credential, error) {
+		reg = mapHostname(reg)
```

Comment:
There are 3 concepts in the docker: registry name, hostname, and credential key.

Let's do some experiments.
```console
$ docker login docker.io -u <username> -p <password>
Login Succeeded
$ cat ~/.docker/config.json
{
        "auths": {
                "https://index.docker.io/v1/": {
                        "auth": "dummy/auth=="
                }
        }
}
$ rm ~/.docker/config.json
$ docker login registry-1.docker.io -u <username> -p <password>
Login Succeeded
$ cat ~/.docker/config.json
{
        "auths": {
                "registry-1.docker.io": {
                        "auth": "dummy/auth=="
                }
        }
}
$ rm ~/.docker/config.json
$ docker login index.docker.io -u <username> -p <password>
Login Succeeded
$ cat ~/.docker/config.json
{
        "auths": {
                "index.docker.io": {
                        "auth": "dummy/auth=="
                }
        }
}
$ rm ~/.docker/config.json
$ docker login https://index.docker.io/v1/ -u <username> -p <password>
Login Succeeded
$ cat ~/.docker/config.json
{
        "auths": {
                "https://index.docker.io/v1/": {
                        "auth": "dummy/auth=="
                }
        }
}
```

As you can observe, 
- The registry name `docker.io` maps to hostname `registry-1.docker.io` where the credential key is `https://index.docker.io/v1/`
- The registry name `registry-1.docker.io` maps to hostname `registry-1.docker.io` where the credential key is `registry-1.docker.io`.

Therefore, `mapHostname()` does not do correct mapping and need to be fixed.

/cc @Wwwsylvia

Reply:
This is very tricky. Thanks @shizhMSFT for the explanation.